### PR TITLE
[refactoring] Refactor Resource message in protobuf to add metric dimension in resource object

### PR DESCRIPTION
**Pull request by rguo-aws**
*Issue #, if available:*

*Description of changes:*
1. Add metric dimension in the resource object defined in protobuf. Now the protobuf resource message can contain "Resource Type" + "Metric". 
i.e. the new resource can have same resource type but with different metric dimension:
"write threadpool" + "queue rejection"  or "write threadpool" + "queue latency"

2. Define the resource object as static variable and put them all into the same ResourceUtil.java so we don't need to re-define them in each RCA / UTs.  

3. remove the oneof structure in resource message as oneof is difficult to be persisted in SQL and de-serialized from SQL. 

*Tests:*
tested on docker

*Code coverage percentage for this patch:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    return resource.getResource().getValueDescriptor().getOptions()
+          .getExtension(PANetworking.additionalFields).getName();
+  }
+
+  /**
+   * Read the resourceType unit type from the ResourceType object
+   * @param resource grpc ResourceType object
+   * @return resource unit type
+   */
+  public static String getResourceTypeUnit(Resource resource) {
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java:71]: The function name here `getResourceTypeUnit` doesn't convey that it is returning resourceMetric information. Consider renaming it.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/HotResourceSummary.java [hunk 2]
```diff
@@ -223,14 +221,14 @@ public String getTableName() {
   public JsonElement toJson() {
     JsonObject summaryObj = new JsonObject();
     summaryObj.addProperty(SQL_SCHEMA_CONSTANTS.RESOURCE_TYPE_COL_NAME,
-        ResourceTypeUtil.getResourceTypeName(this.resourceType));
+        ResourceUtil.getResourceTypeName(this.resource));
+    summaryObj.addProperty(SQL_SCHEMA_CONSTANTS.RESOURCE_METRIC_COL_NAME,
+        ResourceUtil.getResourceTypeUnit(this.resource));
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/HotResourceSummary.java:226]: There is a disconnect between column name and the getter function name. We should rename one of the two.

File: src/main/proto/inter_node_rpc_service.proto [hunk 3]
```diff
@@ -48,40 +48,50 @@ message ResourceContextMessage {
...
 /*
   message for resource type Enum
 */
-message ResourceType {
-    oneof resource_type_oneof {
-      JvmEnum JVM = 1;
-      HardwareEnum hardware_resource_type = 2;
-      ThreadPoolEnum thread_pool = 3;
-    }
+message Resource {
```
Comment (khushbr) [inline src/main/proto/inter_node_rpc_service.proto:92]: We should rename either `message Resource` or the `ResourceEnum resource`. `resourceSummary.getResource().getResource()` isn't very readable.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java [hunk 4]
```diff
@@ -18,12 +18,13 @@
...
 import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics.GCType.TOT_YOUNG_GC;
 import static com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics.HeapDimension.MEM_TYPE;
+import static com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.ResourceUtil.YOUNG_GEN_PROMOTION_RATE;
 
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.PerformanceAnalyzerApp;
-import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.JvmEnum;
-import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceType;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.MetricEnum;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.Resource;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceEnum;
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:26]: Unused imports ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java [hunk 5]
```diff
@@ -15,8 +15,9 @@
 
 package com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.store.rca.hotshard;
 
-import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.HardwareEnum;
-import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceType;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.MetricEnum;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.Resource;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceEnum;
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java:20]: Unused imports ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java [hunk 6]
```diff
@@ -138,7 +140,7 @@ private double getThresholdValue(Map<NodeShardKey, Double> perIndexShardInfo, do
      *
      */
     private void findHotShardAndCreateSummary(Table<String, NodeShardKey, Double> resourceInfoTable, double thresholdInPercentage,
-                                              List<HotResourceSummary> hotResourceSummaryList, ResourceType resourceType) {
+                                              List<HotResourceSummary> hotResourceSummaryList, Resource resourceType) {
```
Comment (khushbr) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java:143]: let's also rename the `resourceType` here.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/HotResourceSummary.java [hunk 7]
```diff
@@ -166,14 +164,14 @@ public static HotResourceSummary buildHotResourceSummaryFromMessage(HotResourceS
   @Override
   public String toString() {
     return new StringBuilder()
-        .append(ResourceTypeUtil.getResourceTypeName(this.resourceType))
+        .append(ResourceUtil.getResourceTypeName(resource))
+        .append(" ")
+        .append(ResourceUtil.getResourceTypeUnit(resource))
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/HotResourceSummary.java:169]: Done, renamed to theResourceMetricName

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.MetricEnum;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.PANetworking;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.Resource;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceEnum;
+
+/**
+ * A utility class to parse and build grpc ResourceType
+ */
+public class ResourceUtil {
+
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java:28]: done, added a javadoc to explain the Resource message and its enums

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java (hunk 1 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/ResourceUtil.java:71]: done, renamed to getResourceMetricName

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java (hunk 4 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotheap/HighHeapUsageYoungGenRca.java:26]: thanks for pointing this out. removed.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java (hunk 5 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java:20]: removed

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java (hunk 6 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/hotshard/HotShardClusterRca.java:143]: done

**Comment by khushbr** [comment]
Thank you for taking up these changes, Ruizhen.

**Pull request merged**